corehq/apps/app_manager/app_schemas/case_properties.py

```diff
@@ -18,130 +20,402 @@
+        all_case_updates = defaultdict(set)
+        for form in self._get_relevant_forms():
+            for case_type, case_properties in form.get_all_case_updates().items():
+                all_case_updates[case_type].update(case_properties)
+        return all_case_updates
 
     @memoized
-    def get_parent_types_and_contributed_properties(self, case_type, include_shared_properties=True):
-        parent_types = set()
-        case_properties = set()
-        forms_info = self.forms_info
-        if self.app.case_sharing and include_shared_properties:
-            forms_info += self.case_sharing_app_forms_info
-
-        for m_case_type, form in forms_info:
-            p_types, c_props = form.get_parent_types_and_contributed_properties(m_case_type, case_type)
-            parent_types.update(p_types)
-            case_properties.update(c_props)
-        return parent_types, case_properties
+    def _get_data_dictionary_properties_by_case_type(self):
+        return {
+            case_type: {prop.name for prop in props} for case_type, props in groupby(
+                CaseProperty.objects
+                .filter(case_type__domain=self.app.domain, deprecated=False)
+                .order_by('case_type__name'),
+                key=attrgetter('case_type.name')
+            )
+        }
+
+    def get_case_relationships_for_case_type(self, case_type):
+        """
+        Like get_case_relationships, but filters down to a single case type
+
+        :param case_type: case type to get relationships for
+        :return: set of (parent, relationship) for this case. See get_case_relationships below.
+
+        """
+        return self.get_case_relationships()[case_type]
+
+    @memoized
+    def get_case_relationships(self):
+        """
+        Returns a `defaultdict(set)` of case_type => set of (parent_type, relationship)
+
+        where relationship is almost always "parent" (but can technically be any word-string).
+
+        This based on all case-subcase relationships appearing in all relevant forms.
+
+        """
+        case_relationships_by_child_type = defaultdict(set)
+
+        for form in self._get_relevant_forms():
+            for case_type, case_relationships in form.get_contributed_case_relationships().items():
+                case_relationships_by_child_type[case_type].update(case_relationships)
+        return case_relationships_by_child_type
 
     def get_parent_types(self, case_type):
-        parent_types, _ = \
-            self.get_parent_types_and_contributed_properties(case_type)
+        """
+        Get a list of all possible parent types for a case type
+
+        Unlike `get_case_relationships`, it doesn't include the `relationship`.
+        That is because it is almost always "parent". Presumably, if this method is used
+        without care, it will cause bugs in the rarer case when relationship is *not* "parent".
+        # todo audit usage to see if such bugs can occur
```

Comment:
I don't want to mess with that actually because I think the one place that does use this may actually use it correctly? As in, it just uses it to figure out what all of the other "relevant" case types are, but doesn't actually assume that they have relationship parent

Reply:
👍 that makes sense